solutions/apio-10-commando.mdx

```diff
@@ -0,0 +1,115 @@
+
+Evaluate the first equation with our new value of $x = \texttt{pre}(i) - \texttt{pre}(j)$
+- $\texttt{dp}(i) = \max_{1 \leq j < i} \{\texttt{dp}[j] + ax^2+bx+c\}$
+- $\texttt{dp}(i) = \max_{1 \leq j < i} \{\texttt{dp}[j] + a[\texttt{pre}(i) - \texttt{pre}(j)]^2+b[\texttt{pre}(i) - \texttt{pre}(j)]+c\}$
+- $\texttt{dp}(i) = \max_{1 \leq j < i} \{\texttt{dp}[j] + [a\cdot\texttt{pre}(i)^2 - 2a\cdot\texttt{pre}(i)\texttt{pre}(j) + a\cdot\texttt{pre}(j)^2] + b\cdot\texttt{pre}(i)-b\cdot\texttt{pre}(j)+c\}$
+
+We will factor terms that do not depend on $j$ outside of the $\max$ statement.
+
+$$\texttt{dp}(i) = \max_{1 \leq j < i} \{\texttt{dp}[j] - 2a\cdot\texttt{pre}(i)\texttt{pre}(j) + a\cdot\texttt{pre}(j)^2 - b\cdot\texttt{pre}(j)\} + a\cdot\texttt{pre}(i)^2 + b\cdot\texttt{pre}(i) + c$$
+
+We will group terms that depend only on $j$ with $\texttt{dp}(j)$.
+
+$$\texttt{dp}(i) = \max_{1 \leq j < i} \{(\texttt{dp}[j] - a\cdot\texttt{pre}(j)^2 - b\cdot\texttt{pre}(j)) - 2a\cdot\texttt{pre}(i)\texttt{pre}(j)\} + a\cdot\texttt{pre}(i)^2 + b\cdot\texttt{pre}(i) + c$$
+
+To efficiently handle transitions in $\mathcal{O}(\log n)$ or $\mathcal{O}(1)$, we will use the convex hull optimization.
+
+**Time Complexity:** $\mathcal{O}(n \log n)$
+
+# Implementation 
+
+<LanguageSection>
+
+<CPPSection>
+  
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using ll = long long;
+
+inline namespace _LineContainer {
+    bool _Line_Comp_State; 
+    struct Line {
+        mutable long long k, m, p;
+        bool operator<(const Line& o) const {
+            return _Line_Comp_State ? p < o.p : k < o.k;
+        }
+    };
+    
+    struct LineContainer : multiset<Line> {
+        const long long inf = LLONG_MAX;
+ 
+        inline long long div(long long a, long long b) { return a / b - ((a ^ b) < 0 && a % b); }
+ 
+        bool isect(iterator x, iterator y) {
+            if (y == end()) { 
+                x->p = inf; 
+                return false; 
+            }
+            if (x->k == y->k) 
+                x->p = x->m > y->m ? inf : -inf;
+            else 
+                x->p = div(y->m - x->m, x->k - y->k);
+            return x->p >= y->p;
+        }
+
+        /**
+         * @param k Slope
+         * @param m Y-Intercept
+         */ 
+        void add(long long k, long long m) {
+            auto z = insert({k, m, 0}), y = z++, x = y;
+            while (isect(y, z)) 
+                z = erase(z);
+            if (x != begin() && isect(--x, y)) 
+                isect(x, y = erase(y));
+            while ((y = x) != begin() && (--x)->p >= y->p)
+                isect(x, erase(y));
+        }
+        
+        long long query(long long x) {
+            assert(!empty());
+            _Line_Comp_State = 1; 
+            auto l = *lower_bound({0,0,x}); 
+            _Line_Comp_State = 0;
+            return l.k * x + l.m;
+        }
+    };
+}
+
+ll N, A, B, C;
+LineContainer lc;
```

Comment:
> is N A B and C do do they have to be global

yes, if it's in the main function it uses too much stack memory or smth and it segfaults

Reply:
they're
they're literally long long's
how much memory can they take up what even

Reply:
Oh I meant the LineContainer
but tbh does it matter? I think global variables look nicer (for CP) and are more utilitarian.

Reply:
do i really have to tell you why global variables are awful

Reply:
do tell

Reply:
https://stackoverflow.com/questions/484635/are-global-variables-bad
and you're literally only using them in `main`
putting them in `main` makes it more obvious that only `main` is using them

Reply:
This question is for general programming, not CP. 
> The problem with global variables is that since every function has access to these, it becomes increasingly hard to figure out which functions actually read and write these variables.

In this case as with most cases, I do want every function to access N, A, B, C (even if it is only the main function in this specific circumstance)

> I've seen programs that seem to double the size of the code by passing an enormous number of parameters around simply to avoid the evil of global variables. In the end, using globals would have made the program clearer to those reading it. By mindlessly adhering to the word of the rule, the original programmer had failed the intent of the rule.
So, yes, globals are often bad. But if you feel that in the end, the intent of the programmer is made clearer by the use of global variables, then go ahead. However, remember the drop in clarity that automatically ensues when you force someone to access a second piece of code (the globals) to understand how the first piece works.

Reply:
>  I do want every function to access

no you don't?
i don't see linecontainer accessing it, so...

Reply:
LineConatiner is just a template, it's copy pastable. 

That's why N, A, B, C are declared under it (so LineContainer can't access it anyways)

Reply:
so...why are they global now? i don't feel like they make the intent clear or whatever
you say you want every function to access it, but where are the supposed functions?

Reply:
I meant it to be applied to a general case. In this specific example, there happens to only be one function which is the `main`. In general, global variables have the added benefit of initializing each type to its default value. 

You'll notice that I declared `P` and `dp` as variables local to the main because they belong to the processes and algorithms being run there (in that they would be useless after said processes terminate). 

Although none of the above statements apply to the current problem, consistency is better than formatting being evaluated on a case-by-case basis. 

Also, on an unrelated side note, why did you move the declaration of `a` outside of the loop? If you care so much about scope, then `int a` belongs right before the line inputting it, as it is only used as a temporary variable for reading in the input.

Reply:
what do you mean i didn't even commit to this pr bruh

Reply:
also mods already said consistency doesn't matter that much for unimportant things, and this is probably one of those "don't really care things"

Reply:
> also mods already said consistency doesn't matter that much for unimportant things, and this is probably one of those "don't really care things"

I feel like as long as the variable names are relatively clear (which they are), it should be fine.

Reply:
how are they clear
i have no idea what those mean
anyways i'm fine with the LC being global but these have literally no reason to be

Reply:
they're constants provided in the problem statement?

Reply:
ok sure
but even then
how does making these global help you make your code clearer

Reply:
Because the declarations aren't mixed with the code itself

Reply:
the variables are literally part of the code???

Reply:
yes variables are part of codes, that's what they do

Reply:
so why not just make that clear and put them in main